cns/restserver/api.go

```diff
@@ -899,6 +902,99 @@ func (service *HTTPRestService) getNetworkContainerByOrchestratorContext(w http.
 	logger.Response(service.Name, getNetworkContainerResponse, returnCode, err)
 }
 
+// getOrRefreshNetworkContainers is to check whether refresh association is needed.
+// If received  "GET": return all NCs in CNS's state file to DNC for follow-up check
+// If received "POST": Store all the NCs (from the request body that client sent) into CNS's state file
+func (service *HTTPRestService) getOrRefreshNetworkContainers(w http.ResponseWriter, r *http.Request) {
+	logger.Printf("[Azure CNS] getOrRefreshNetworkContainers")
+
+	switch r.Method {
+	case "GET":
+		// return all NCs in CNS
+		networkContainersDetails := make(map[string]cns.ContainerStatus)
+		for ncID := range service.state.ContainerStatus {
+			networkContainersDetails[ncID] = storeNcStatus(service.state.ContainerStatus[ncID])
+		}
+
+		response := cns.GetOrRefreshNetworkContainersResponse{
+			NetworkContainersDetails: networkContainersDetails,
+			Response: cns.Response{
+				ReturnCode: types.Success,
+			},
+		}
+		err := service.Listener.Encode(w, &response)
+		logger.Response(service.Name, response, response.Response.ReturnCode, err)
+
+	case "POST":
+		var req cns.GetOrRefreshNetworkContainersRequest
+		err := service.Listener.Decode(w, r, &req)
+		logger.Request(service.Name, &req, err)
+		if err != nil {
+			response := cns.GetOrRefreshNetworkContainersResponse{
+				Response: cns.Response{
+					ReturnCode: types.InvalidRequest,
+					Message:    fmt.Sprintf("[Azure CNS] Create Network Container failed with error: %s", err.Error()),
+				},
+			}
+			err = service.Listener.Encode(w, &response)
+			logger.Response(service.Name, response, response.Response.ReturnCode, err)
+			return
+		}
+
+		returnCode := types.Success
+		returnMessage := ""
+		for i := 0; i <= len(req.CreateNetworkContainerRequests); i++ {
+			body, encodeErr := json.Marshal(req.CreateNetworkContainerRequests[i])
+			if encodeErr != nil {
+				returnCode = types.InvalidRequest
+				returnMessage = fmt.Sprintf("[Azure CNS] Create Network Container failed with error: %s", err.Error())
+				break
+			}
+			createNcRequest, requestErr := http.NewRequestWithContext(context.Background(), http.MethodPost, "", bytes.NewBuffer(body))
+			if requestErr != nil {
+				returnCode = types.InvalidRequest
+				returnMessage = fmt.Sprintf("[Azure CNS] Create Network Container failed with error: %s", err.Error())
+				break
+			}
+			createNcRequest.Header.Set(common.ContentType, common.JsonContent)
+			writer := httptest.NewRecorder()
+			service.createOrUpdateNetworkContainer(writer, createNcRequest)
```

Comment:
this does not look right. we shouldn't need to be creating `http.Requests` for this functionality.

Reply:
Fixed.